src/mapstorage/XmlMapStorage.cpp

```diff
@@ -27,11 +35,160 @@
+// list know enum types
+#define X_FOREACH_TYPE_ENUM(X) \
+    X(RoomAlignEnum) \
+    X(DoorFlagEnum) \
+    X(ExitFlagEnum) \
+    X(RoomLightEnum) \
+    X(RoomLoadFlagEnum) \
+    X(InfoMarkClassEnum) \
+    X(InfoMarkTypeEnum) \
+    X(RoomMobFlagEnum) \
+    X(RoomPortableEnum) \
+    X(RoomRidableEnum) \
+    X(RoomSundeathEnum) \
+    X(RoomTerrainEnum) \
+    X(Type)
+
+enum class XmlMapStorage::Type : uint {
+#define DECL(X) X,
+    X_FOREACH_TYPE_ENUM(DECL)
+#undef DECL
+};
+
+// ---------------------------- XmlMapStorage::Converter -----------------------
+class XmlMapStorage::Converter final
+{
+public:
+    Converter();
+    ~Converter() = default;
+
+    // parse string containing a signed or unsigned number.
+    // sets fail = true only in case of errors, otherwise fail is not modified
+    template<typename T>
+    static T toInteger(const QStringView str, bool &fail)
+    {
+        bool ok = false;
+        const T ret = to_integer<T>(as_u16string_view(str), ok);
+        if (!ok) {
+            fail = true;
+        }
+        return ret;
+    }
+
+    // parse string containing the name of an enum.
+    // sets fail = true only in case of errors, otherwise fail is not modified
+    template<typename ENUM>
+    ENUM toEnum(const QStringView str, bool &fail) const
+    {
+        static_assert(std::is_enum<ENUM>::value, "template type ENUM must be an enumeration");
+        return ENUM(stringToEnum(enumToType(ENUM(0)), str, fail));
+    }
+
+    template<typename ENUM>
+    const QString &toString(ENUM val) const
+    {
+        static_assert(std::is_enum<ENUM>::value, "template type ENUM must be an enumeration");
+        return enumToString(enumToType(val), uint(val));
+    }
+
+private:
+    // define a bunch of methods
+    //   static constexpr Type enumToType(RoomAlignEnum) { return Type::RoomAlignEnum; }
+    //   static constexpr Type enumToType(DoorFlagEnum)  { return Type::DoorFlagEnum;  }
+    //   ...
+    // converting an enumeration type to its corresponding Type value,
+    // which can be used as argument in enumToString() and stringToEnum()
+#define DECL(X) \
+    static constexpr Type enumToType(X) { return Type::X; }
+    X_FOREACH_TYPE_ENUM(DECL)
+#undef DECL
+
+    uint stringToEnum(Type type, const QStringView str, bool &fail) const;
+    const QString &enumToString(Type type, uint val) const;
+
+    std::vector<std::vector<QString>> enumToStrings;
+    std::vector<QHash<QStringView, uint>> stringToEnums;
+    QString empty;
+};
+
+XmlMapStorage::Converter::Converter()
+    : enumToStrings{
+#define DECL(X) /* */ {#X},
+#define DECL_(X, ...) {#X},
+        {X_FOREACH_RoomAlignEnum(DECL)},
+        {X_FOREACH_DOOR_FLAG(DECL_)},
+        {X_FOREACH_EXIT_FLAG(DECL_)},
+        {X_FOREACH_RoomLightEnum(DECL)},
+        {X_FOREACH_ROOM_LOAD_FLAG(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
```

Comment:
We accidentally declared this type twice

Reply:
Thanks for catching this. I'll fix it and add a check to ensure that `enumToStrings` initialization remains in sync with `X_FOREACH_TYPE_ENUM`

Reply:
Fixed. the last line above is now `{X_FOREACH_INFOMARK_TYPE(DECL)},`